File: Github.lua
local req      = require("requests")
local json     = require("rapidjson")
local argparse = require("argparse")

-- Base url for requests
local BASE_URL = "https://api.github.com/users/"

---@class GthUser @Class for a github user basic info
---@field url       string (private) Url to make requests, for internal usage
---@field json      table  (private) Json response of the request
---@field name      string (public)  User name as shown in the github profile
---@field bio       string (public)  User bio as shown in the github profile
---@field link      string (public)  User profile url
---@field repos     table  (public)  Public user repositories
---@field gists     table  (public)  Public user gists
---@field followers table  (public)  User followers
---@field following table  (public)  User following
local GthUser = {}

--- Constructor for GthUser class
---@param username string The github username to get info
---@return GthUser
function GthUser:init(username)
  assert(
    type(username) == "string",
    "Bad argument for 'init', string expected, got " .. type(username)
  )

  self.url = BASE_URL .. username
  local res = req.get(self.url)
  self.json = json.decode(res.text)

  self.name = self.json['name']
  self.bio  = self.json['bio']
  self.link = self.json['html_url']

  self.repos = { count = self.json['public_repos'], arr = {} }
  self.gists = { count = self.json['public_gists'], arr = {} }
  self.followers = { count = self.json['followers'],    arr = {} }
  self.following = { count = self.json['following'],    arr = {} }

  return self
end

--- Gets the requested data (repos, gists, followers and following)
---@param thing string The data to request
---@return nil
function GthUser:fetch(thing)
  assert(
    type(thing) == "string",
    "Bad argument for 'fetch', string expected, got " .. type(thing)
  )

  local url = self.url .. '/' .. thing
  local res = req.get(url)
  local arr = json.decode(res.text)

  if thing == "repos" then
    for i, v in ipairs(arr) do
      self.repos.arr[i] = v["name"]
    end
  elseif thing == "gists" then
    for i, v in ipairs(arr) do
      self.gists.arr[i] = v["description"]
    end
  elseif thing == "followers" then
    for i, v in ipairs(arr) do
      self.followers.arr[i] = v["login"]
    end
  elseif thing == "following" then
    for i, v in ipairs(arr) do
      self.following.arr[i] = v["login"]
    end
  else
    return error("Unsupported endpoint: " .. thing)
  end
end

local opts = argparse({
  name = "Github.lua",
  description = "Simple example of the Github's REST API",
  epilog = "Check out https://github.com/M1que4s/Github-REST-API-Example"
})

opts:argument("usernames", "One or more usernames"):args("+")
opts:flag("-f --followers", "Shows user followers", false)
opts:flag("-F --following", "Shows user following", false)
opts:flag("-r --repos", "Shows user repos", false)
opts:flag("-g --gists", "Shows user gists", false)

local args = opts:parse(arg)

for _, user in ipairs(args.usernames) do
  local user = GthUser:init(user)

  print("name: " .. user.name)
  print("bio: " .. user.bio)
  print("link: " .. user.link)

  print("Public repos: " .. user.repos.count)
  if args.r or args.repos then
    user:fetch("repos")

    for i, v in ipairs(user.repos.arr) do
      print(("| %02d %s: %s"):format(i, v.name, v.Desc))
    end
  end

  print("Public gists: " .. user.gists.count)
  if args.g or args.gists then
    user:fetch("gists")

    for i, v in ipairs(user.gists.arr) do
      print(("| %02d. %s"):format(i, v))
    end
  end

  print("Followers: " .. user.followers.count)
  if args.f or args.followers then
    user:fetch("followers")

    for _, v in ipairs(user.followers.arr) do
      print("| @" .. v)
    end
  end

  print("Following: " .. user.following.count)
  if args.F or args.following then
    user:fetch("following")

    for _, v in ipairs(user.following.arr) do
      print("| @" .. v)
    end
  end

  print()
end

File: Github.nim
import std/httpclient
import std/strformat
import std/parseopt
import std/json
from std/os import commandLineParams

const BASE_URL = "https://api.github.com/users/"

type
  GthUserItem* = object
    count*: int
    arr*: seq[string]

  GthUser* = object
    # Private
    obj: JsonNode
    url: string
    # Public
    name*: string
    bio*: string
    link*: string
    repos*: GthUserItem
    gists*: GthUserItem
    followers*: GthUserItem
    following*: GthUserItem

proc init*(self: var GthUser; username: string) =
  self.url = BASE_URL & username
    
  var res  = newHttpClient().getContent(self.url)
  self.obj = parseJson(res)

  self.name = self.obj["name"].getStr()
  self.bio  = self.obj["bio"].getStr()
  self.link = self.obj["html_url"].getStr()

  self.repos = GthUserItem(count: self.obj["public_repos"].getInt())
  self.gists = GthUserItem(count: self.obj["public_gists"].getInt())
  self.followers = GthUserItem(count: self.obj["followers"].getInt())
  self.following = GthUserItem(count: self.obj["following"].getInt())

proc fetch*(self: var GthUser; thing: string) =
  var
    url = self.url & '/' & thing
    res = newHttpClient().getContent(url)
    arr = parseJson(res)
  
  case thing
    of "repos":
      for v in arr.items():
        self.repos.arr.add(v["name"].getStr())
    of "gists":
      for v in arr.items():
        self.gists.arr.add(v["description"].getStr())
    of "followers":
      for v in arr.items():
        self.followers.arr.add(v["login"].getStr())
    of "following":
      for v in arr.items():
        self.following.arr.add(v["login"].getStr())
    else:
      echo "Unsupported endpoint: ", thing

var
  cmdline = commandLineParams()
  args: seq[string]
  opts: tuple[
    repos: bool,
    gists: bool,
    followers: bool,
    following: bool
  ]

for kind, name, value in getopt(cmdline):
  case kind:
    of cmdArgument: args.add(name)
    of cmdShortOption, cmdLongOption:
      case name:
        of "r", "repos": opts.repos = true
        of "g", "gists": opts.gists = true
        of "f", "followers": opts.followers = true
        of "F", "following": opts.following = true
        else: continue
    of cmdEnd: break

if args.len() == 0:
  echo "No arguments, nothing to do."
else:
  for username in args:
    var user = GthUser()
    user.init(username)

    echo "Name: ", user.name
    echo "Bio: ", user.bio
    echo "Link: ", user.link

    echo "Public repos: ", user.repos.count
    if opts.repos:
      user.fetch("repos")

      for i, v in user.repos.arr:
        echo fmt"| {(i + 1):03}. {v}"
    
    echo "Public gists: ", user.gists.count
    if opts.gists:
      user.fetch("gists")

      for i, v in user.gists.arr:
        echo fmt"| {(i + 1):03}. {v}"
    
    echo "Public followers: ", user.followers.count
    if opts.followers:
      user.fetch("followers")

      for v in user.followers.arr:
        echo fmt"| @{v}"
    
    echo "Public following: ", user.following.count
    if opts.following:
      user.fetch("following")

      for v in user.following.arr:
        echo fmt"| @{v}"
    
    echo ""